app/Jobs/RockTheVote/ImportRockTheVoteRecord.php

```diff
@@ -76,7 +76,14 @@ public function handle(Rogue $rogue)
             return;
         }
 
-        $this->updateUserIfChanged($user);
+        $updateUserPayload = $this->getUpdateUserPayload($user);
```

Comment:
I was struggling a bit with following some of the code in this PR, and I think it mostly has to do with the naming 😬 `$updateUserPayload` sounds like it's a boolean; deciding whether it should or shouldn't update the payload.

Maybe rename the variable to `$updatedUserData`?

Reply:
I'm not too keen on calling it `updated` when it hasn't been updated yet, this is the payload we're going to use for making a request to update the user. I don't agree that `updateUserPayload` sounds like a boolean -- I thought the `Payload` would indicate it's an object (or an associate array in PHP land). Maybe something like `userFieldValuesToChange`?

Reply:
Ah, ok. See part of the confusion with knowing if it's updated yet is because the naming sounded odd to me, along with `getUpdateUserPayload()`. Which is why I was stuck on it trying to decipher what was going on.

Ya, payload does indicate an object/array but "update user payload?" sounds like a question to me, needing to decide whether the payload should or shouldn't be updated; because it's not obvious to read it as "updateUser() payload" as in the payload for the `updateUser()` method.

Reply:
The previous method of `updateUserIfChanged()` was more straightforward to me, because it indicates that if the user info changed, lets go ahead and update it. Is there a reason we can't keep that method as it was named and then rework the logic in there to deal with the feature flag etc?

Reply:
That's fair enough.  Would just switching the order to `userUpdatePayload` work? Then the other function could be named `getUserUpdatePayload`

Reply:
@aaronschachter not sure if you saw my comment about `updateUserIfChanged()` since I think we submitted comments at the exact same time!

![Muscle High Five](https://user-images.githubusercontent.com/105849/79022472-02078680-7b4c-11ea-9348-5d1b049a3160.gif)

Reply:
Just popping in from #notify-applications… 🤓 I can definitely see how this method could be read two entirely different ways. It's like optical illusions where it can look like something's spinning clockwise or counter-clockwise just depending on how it "clicks" in your brain.

Renaming this to `userUpdatePayload` and `getUserUpdatePayload` definitely feels like an improvement.

I also like @weerd's idea of keeping these in one method to sidestep the naming concerns altogether (since as far as I know this method isn't used anywhere else, right?). That might be a little easier to read, you can step through all the logic that goes into updating a user (or not) sequentially.

Reply:
The discussion over in https://github.com/DoSomething/chompy/pull/148#discussion_r401756658 was one prompt that led me to refactor `updateUserIfChanged` into smaller function, to test each one. The part 2 of this PR would be to change a user's `sms_status` and `sms_subscription_topics` per row values, which adds even more permutations to test. For those, I was thinking of another smaller function like `getUserSmsStatusAndSubscriptionTopicsPayload` to test those variations vs all of them in one method.

It does make testing each smaller function easier, which is why these functions are public, so we can just test each payload-getter. For sure, it could be refactored to test all as `updateUserIfChanged` though... I can try that first to see if there were any gotchas with going that route, vs renaming multiple times/functions.

Reply:
Could you maybe do a bit of both... keep `updateUserIfChanged()` since it makes sense and then in there it does all the calls to the smaller functions. Those smaller functions could still be tested independently, and the `updateUserIfChanged()` is just a convenience method that kind of collects all that logic into an easier to call place. Does that make sense? Maybe this is already what you mean tho 😅

Reply:
Ah, I thought it was cleaner to break out of each function when we have smaller functions that return payloads to construct -- e.g. There's no mobile here? Ok return an empty array -- vs a bunch of if/else -- also per https://github.com/DoSomething/chompy/pull/148#discussion_r401793957 review.

Reply:
> `updateUserIfChanged()` is just a convenience method that kind of collects all that logic into an easier to call place.

That's what the `getUserUpdatePayload` is doing -- collecting all the logic. All `updateUserIfChanged` would look like is
```
if ($data = getUserUpdatePayload($user)) {
   gateway->northstar->updateUser($data)
}
```
which is where it seemed straight forward enough to have this logic in the `handle` as is in this PR vs a separate `updateUserIfChanged`. Per my last comment, being able to exit out of each payload building function was what felt cleaner than nested if/else statements and calling Northstar update user function at the end.

Reply:
I'm thinking something like this:

```php
public function updateUserIfChanged($user)
{
  $payload = [];

  $payload = array_merge($payload, methodThatReturnsArrayIfMobileChanged());

  $payload = array_merge($payload, methodThatReturnsArrayIfStatusChanged());

  $payload = array_merge($payload, methodThatReturnsArrayIfSomeOtherFieldChanged());

  if(count($payload)) {
    // send request to update the user's information
  } else {
    // no update needed
  }
}
```

Pseudo-ish code, but hopefully that makes sense. Each separate method can still be tested separately. This also still alleviates the amount of logic in just the `updateUserIfChanged()` method, but spread out into more specific, smaller functions.

Reply:
Makes sense! Made this change, thanks for the example.